### ISSUE TITLE
Header._keyword_from_index() selects wrong keyword for negative indices

### ISSUE THREAD

**Issue by rsiverd** [astropy/astropy#5827]
I have a script that loads an image and header from a FITS file, appends new keywords to the header, fiddles with the image, and saves the result (with augmented header) to a new file. I recently noticed that this script fails for Rice-compressed FITS images but works fine with the same image if decompressed in advance with funpack.

Basic idea:
```
data, header = pf.getdata(input_file, header=True)
while (len(header) < 1000):
    header.append()    # pad with grow room
....
# Append stats to header:
header.append(("Q1_OSAVG", avg, "[adu] quadrant 1 overscan mean"))
header.append(("Q1_OSSTD", avg, "[adu] quadrant 1 overscan stddev"))
header.append(("Q1_OSMED, avg, "[adu] quadrant 1 overscan median"))  # this fails

```
The [traceback](https://github.com/astropy/astropy/files/784647/traceback.txt) (attached) indicates that an IndexError occurs when appending a new keyword to the CompImageHeader. Investigating the cause I found that the _keyword_indices() routine both (a) selects the wrong keyword and (b) mis-counts the number of repeats. 

The following commands in ipython (also attached [index_error_ipython.txt](https://github.com/astropy/astropy/files/784714/index_error_ipython.txt)) should illustrate the problem:
<pre>
In [1]: import astropy.io.fits as pf

In [2]: hdr = pf.Header([ ('MOONFRAC', 0.604829), ('MOONDIST', 82.5171118), ('MOONALT', 7.5409339), ('SUNDIST', 56.3424603), ('SUNALT', 33.6685322),])

In [3]: hdr
Out[3]: 
MOONFRAC=   0.6048289999999999                                                  
MOONDIST=           82.5171118                                                  
MOONALT =            7.5409339                                                  
SUNDIST =           56.3424603                                                  
SUNALT  =           33.6685322                                                  

In [4]: hdr[-1]
Out[4]: 33.6685322

In [5]: hdr._keyword_from_index(-1)
Out[5]: ('SUNDIST', 0)
</pre>

A quick value inspection at the time of error in my script illustrates the incorrect repeat count:
```
>>> print len(header._keyword_indices[''])
774

>>> print "blanks: %d" % len([x for x in header.keys() if x==''])
blanks: 774

# Problem:
>>> print header._keyword_from_index(-1)  # should return 773 repeats (774 total present)
('', 772)
```

Looking at the header.py source code shows there is an off-by-one error in the index (line 1697):
```
currently -->    idx += len(self._cards) - 1
should be -->    idx += len(self._cards)
```
If a list has 10 elements and -1 is the last, the true index should be 9. This error occurs also in the insert() method.

For now, I can use the copy() method to strip out any existing blanks from the FITS file just after load and then append new ones just prior to save. Given the importance of consistent indexing, however, I thought it best to report this problem. 

I spotted this bug using Python 2.7.12, astropy 1.1.2, and numpy 1.10.4 but checked and found the bug above is still present in the master branch on github [here](https://github.com/astropy/astropy/blob/master/astropy/io/fits/header.py).

**Comment by pllim** [comment]
Any chance you can provide the FITS file? Or test it under Python 3?

**Comment by pllim** [comment]
@rsiverd , since you basically laid out how to fix this, would you be interested to submit a PR? If you can't, no problem. But if you can, you can be credited with the fix.

**Comment by rsiverd** [comment]
I'll see if I can get it fixed up today and let you know how it goes.

**Comment by mohanagr** [comment]
@pllim I am able to reproduce this in python3. @rsiverd Trying to fix this.

**Comment by mohanagr** [comment]
Submitted a PR based on what @rsiverd pointed out.

**Comment by pllim** [comment]
Should be fixed in #5866. Thank you very much for suggesting a fix.
